library/core/src/iter/traits/iterator.rs

```diff
@@ -1670,6 +1672,77 @@ pub trait Iterator {
         FromIterator::from_iter(self)
     }
 
+    /// Collects all items from the iterator into an array of a specific size.
+    ///
+    /// If the number of elements inside the iterator is exactly equal to the
+    /// array size, then the array is returned inside `Some`, otherwise `None`
+    /// is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(iter_collect_array)]
+    ///
+    /// let xs = [1, 2, 3];
+    /// let iter = xs.iter().copied();
+    /// let [_a, b, _c] = iter.clone().collect_array().unwrap();
+    /// assert_eq!(b, 2);
+    ///
+    /// match iter.collect_array() {
+    ///     Some([_, _]) => panic!("Didn't expect to see two only elements"),
+    ///     None => (),
+    /// }
+    /// ```
+    #[inline]
+    #[unstable(reason = "new API", issue = "none", feature = "iter_collect_array")]
+    fn collect_array<const N: usize>(self) -> Option<[Self::Item; N]>
```

Comment:
pondering: given that this will potentially collect only a prefix, should it be `&mut self`?  The `Option` is enough to know whether the iterator was exhausted in the call, so one could use `.collect_array::<4>()` in a loop...

Reply:
Ah, I think I misunderstood initially... You want to have "exact N" semantics, but you want `&mut self` at the same time?

This is interesting!

The problem I am seeing here is that, to get exactly N, you'll need to fetch N + 1 elements, so one would be dropped on the floor or stuffed into the `Err` variant somehow.

Reply:
I was thinking that `it.collect_array::<3>()` would be equivalent to `try { [it.next()?, it.next()?, it.next()?] }`.  So you'd get `None` if the length is shorter than the number requested, but if there's more you'd be able to call again to get more of them.

(Maybe with a rename like sfackler mentions -- this is reminding me of `as_chunks` and `array_chunks` on slices, so maybe something like that?)

Reply:
@scottmcm for iterators,  `self` and `&mut self` is the same thing - you can always use `.by_ref()`

Reply:
@WaffleLapkin Conceptually yes.  But in implementation not exactly -- see [all these benchmarks](https://github.com/rust-lang/rust/blob/e9920ef7749d11fc71cc32ca4ba055bcfeaab945/library/core/benches/iter.rs#L146-L148) for example.

And if this doesn't exhaust the iterator, then it should be `&mut self` for consistency with other iterator methods that don't necessarily exhaust the iterator (like `all` and `any`, which are also `&mut self`).